physical/raft/raft.go

```diff
@@ -474,9 +480,55 @@ func (b *RaftBackend) Close() error {
 		return err
 	}
 
+	close(b.statsDoneCh)
+
 	return nil
 }
 
+func (b *RaftBackend) collectMetrics() {
+	if b.metricSink == nil {
+		b.logger.Warn("collectMetrics() called without a metricSink set, aborting")
+		return
+	}
+
+	b.l.Lock()
+	b.statsDoneCh = make(chan struct{})
+	b.l.Unlock()
+
+	go func() {
+		for {
+			select {
+			case <-b.statsDoneCh:
+				return
+			default:
+				b.l.RLock()
+				stats := b.stableStore.(*raftboltdb.BoltStore).Stats()
+				txstats := stats.TxStats
+				b.metricSink.SetGauge([]string{"raft", "bolt_store", "freelist", "free_pages"}, float32(stats.FreePageN))
+				b.metricSink.SetGauge([]string{"raft", "bolt_store", "freelist", "pending_pages"}, float32(stats.PendingPageN))
+				b.metricSink.SetGauge([]string{"raft", "bolt_store", "freelist", "allocated_bytes"}, float32(stats.FreeAlloc))
+				b.metricSink.SetGauge([]string{"raft", "bolt_store", "freelist", "used_bytes"}, float32(stats.FreelistInuse))
+				b.metricSink.SetGauge([]string{"raft", "bolt_store", "transaction", "started_read_transactions"}, float32(stats.TxN))
+				b.metricSink.SetGauge([]string{"raft", "bolt_store", "transaction", "currently_open_read_transactions"}, float32(stats.OpenTxN))
+				b.metricSink.SetGauge([]string{"raft", "bolt_store", "page", "count"}, float32(txstats.PageCount))
+				b.metricSink.SetGauge([]string{"raft", "bolt_store", "page", "bytes_allocated"}, float32(txstats.PageAlloc))
+				b.metricSink.SetGauge([]string{"raft", "bolt_store", "cursor", "count"}, float32(txstats.CursorCount))
+				b.metricSink.SetGauge([]string{"raft", "bolt_store", "node", "count"}, float32(txstats.NodeCount))
+				b.metricSink.SetGauge([]string{"raft", "bolt_store", "node", "dereferences"}, float32(txstats.NodeDeref))
+				b.metricSink.SetGauge([]string{"raft", "bolt_store", "rebalance", "count"}, float32(txstats.Rebalance))
+				b.metricSink.AddSample([]string{"raft", "bolt_store", "rebalance", "time"}, float32(txstats.RebalanceTime))
+				b.metricSink.SetGauge([]string{"raft", "bolt_store", "split", "count"}, float32(txstats.Split))
+				b.metricSink.SetGauge([]string{"raft", "bolt_store", "spill", "count"}, float32(txstats.Spill))
+				b.metricSink.AddSample([]string{"raft", "bolt_store", "spill", "time"}, float32(txstats.SpillTime))
+				b.metricSink.SetGauge([]string{"raft", "bolt_store", "write", "count"}, float32(txstats.Write))
+				b.metricSink.AddSample([]string{"raft", "bolt_store", "write", "time"}, float32(txstats.WriteTime))
+				b.l.RUnlock()
+				time.Sleep(time.Second)
```

Comment:
Should there be some way (not necessarily user-facing) to configure the interval?

Reply:
Sure? I'm open to suggestions.

Reply:
Actually, I retract the suggestion.  What I think I'd prefer would be to have `metricsLoop` call your collectMetrics, which would contain only the loop body here, no goroutine.  That way we have one place in Vault that does the periodic collection, and if we want to adjust the frequency, we can do it for all wholesale.  You could also make metricsSink be an argument to this function, so we don't need the setter.